<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.graph_objects as go
from dotenv import load_dotenv
import os


load_dotenv()

True

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = os.getenv('DBNAME')
USER = os.getenv('USER')
PASSWORD = os.getenv('PASSWORD')
HOST = os.getenv('HOST')
PORT = os.getenv('PORT')

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''
select count(*) from vacancies
'''

In [5]:
# результат запроса
df_3_1 = pd.read_sql_query(query_3_1, connection)
df_3_1

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\4045908245.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_1 = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''
select count(*) from employers
'''

In [7]:
# результат запроса
df_3_2 = pd.read_sql_query(query_3_2, connection)
df_3_2

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\533540996.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_2 = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''
select count(*) from areas
'''

In [9]:
# результат запроса
df_3_3 = pd.read_sql_query(query_3_3, connection)
df_3_3

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\1395976625.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_3 = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''
select count(*) from industries
'''

In [11]:
# результат запроса
df_3_4 = pd.read_sql_query(query_3_4, connection)
df_3_4

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\854197215.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_4 = pd.read_sql_query(query_3_4, connection)


count
0    294

***

## Дополнительное исследование

Определение типов рабочего графика

In [12]:
# текст запроса
query_3_5 = f'''
select distinct schedule from vacancies
'''

df_3_5 = pd.read_sql_query(query_3_5, connection)
df_3_5

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\2540820828.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_5 = pd.read_sql_query(query_3_5, connection)


schedule
0    Сменный график
1    Вахтовый метод
2     Гибкий график
3  Удаленная работа
4       Полный день

Определение типов трудоустройства

In [13]:
# текст запроса
query_3_6 = f'''
select distinct employment from vacancies
'''

df_3_6 = pd.read_sql_query(query_3_6, connection)
df_3_6

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\1998022503.py:6: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_6 = pd.read_sql_query(query_3_6, connection)


employment
0           Стажировка
1  Частичная занятость
2     Проектная работа
3     Полная занятость

Топ 10 сфер деятельности по кол-ву вакансий

In [14]:
# текст запроса
query_3_7 = f'''
select 
    i.name as industry_name,
    count(*) as cnt_vacs
from vacancies v
left join employers e on v.employer_id = e.id
left join employers_industries ei on e.id = ei.employer_id
left join industries i on i.id = ei.industry_id
group by 1
order by 2 desc
limit 10
'''

df_3_7 = pd.read_sql_query(query_3_7, connection)
df_3_7

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\86619181.py:15: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_3_7 = pd.read_sql_query(query_3_7, connection)


industry_name  cnt_vacs
0                Разработка программного обеспечения     12499
1  Системная интеграция,  автоматизации технологи...     11034
2                                               None     10084
3  Интернет-компания (поисковики, платежные систе...      6413
4                                               Банк      2742
5                                    Мобильная связь      1585
6                                 Кадровые агентства      1143
7                              Консалтинговые услуги      1137
8                                 Интернет-провайдер      1095
9                       Розничная сеть (продуктовая)      1034

# Выводы:

В данных присутствует всего 49197 вакансий, 23501 записи о работодателях, данные о 1362 регионах и 294 сферы деятельности работодателей. Примерно две вакансии приходиться на каждого работодателя. Также больше всего вакансий присутствует из сферы разработки программного обеспечения. Наиболее популярными сферами деятельности работодателей являются сферы так или иначе связанные с IT. Это может указывать на ориентированность HH на сферу IT.
На агрегаторе вакансий представлено 5 типов рабочих графиков (Сменный график, Вахтовый метод, Гибкий график, Удаленная работа, Полный день) и 4 типа занятости (Стажировка, Частичная занятость, Проектная работа, Полная занятость).

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [15]:
# текст запроса
query_4_1 = f'''
select 
    count(*) as cnt,
    a.name
from vacancies v
join areas a on v.area_id = a.id
group by a.name
order by 1 desc
'''

In [16]:
# результат запроса
df_4_1 = pd.read_sql_query(query_4_1, connection)
df_4_1

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\620650199.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df_4_1 = pd.read_sql_query(query_4_1, connection)


cnt             name
0    5333           Москва
1    2851  Санкт-Петербург
2    2112            Минск
3    2006      Новосибирск
4    1892           Алматы
..    ...              ...
764     1       Тарко-Сале
765     1    Новоаннинский
766     1            Бирск
767     1           Сасово
768     1             Анна

[769 rows x 2 columns]

Визуализация запроса

In [17]:
fig_bar = px.bar(
    data_frame=df_4_1.nlargest(10, 'cnt'),
    x='name',
    y='cnt',
    color='name',
    text='cnt',
    orientation='v',
    height=500,
    width=1000,
    title='Топ-10 регионов по вакансиям')


fig_bar.show()
fig_bar.write_image('graphs/img1.png')
fig_bar.write_html("html/img1.html")

![Топ-10 регионов по вакансиям'](graphs/img1.png)

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [18]:
# текст запроса
query_4_2 = f'''
select 
    count(*) as cnt
from vacancies v
where v.salary_from is not null or v.salary_to is not null
'''

In [19]:
# результат запроса
df_4_2 = pd.read_sql_query(query_4_2, connection)
df_4_2

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\3746829199.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [20]:
# текст запроса
query_4_3 = f'''
select 
    round(avg(v.salary_from)) as avg_salary_from,
    round(avg(v.salary_to)) as avg_salary_to
from vacancies v
'''

In [21]:
# результат запроса
df_4_3 = pd.read_sql_query(query_4_3, connection)
df_4_3

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\3595259731.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [22]:
# текст запроса
query_4_4 = f'''
select 
    v.schedule,
    v.employment,
    count(*) as cnt
from vacancies v
group by 1, 2
order by 3 desc
'''

In [23]:
# результат запроса
df_4_4 = pd.read_sql_query(query_4_4, connection)
df_4_4

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\2028972612.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



schedule           employment    cnt
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

Визуализация запроса

In [24]:
fig_bar = px.bar(
    data_frame=df_4_4,
    x='schedule',
    color='employment',
    y='cnt',
    barmode='group',
    text='cnt',
    orientation='v',
    title='Кол-во вакансий по группам "тип графика работы"-"тип занятости"')

fig_bar.show()
fig_bar.write_image('graphs/img2.png')
fig_bar.write_html("html/img2.html")

![Кол-во вакансий по группам "тип графика работы"-"тип занятости"](graphs/img2.png)

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [25]:
# текст запроса
query_4_5 = f'''
select 
    v.experience,
    count(*) as cnt  --count считает только не null значения
from vacancies v
group by 1
order by 2
'''

In [26]:
# результат запроса
df_4_5 = pd.read_sql_query(query_4_5, connection)
df_4_5

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\210463167.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



experience    cnt
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

## Дополнительное исследование

Топ-10 самых популярных сочетаний сферы деятельности и требуемого опыта в вакансиях

In [27]:
# текст запроса
query_4_6 = f'''
select 
    i.name as ind_name,
    v.experience,
    count(*) as cnt  --count считает только не null значения
from vacancies v
left join employers e on v.employer_id = e.id
left join employers_industries ei on e.id = ei.employer_id
left join industries i on i.id = ei.industry_id
where i.name is not null
group by 1, 2
order by 3 desc
limit 10
'''

# результат запроса
df_4_6 = pd.read_sql_query(query_4_6, connection)
df_4_6

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\279587693.py:18: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



ind_name          experience   cnt
0                Разработка программного обеспечения  От 1 года до 3 лет  6421
1  Системная интеграция,  автоматизации технологи...  От 1 года до 3 лет  5829
2                Разработка программного обеспечения       От 3 до 6 лет  4526
3  Системная интеграция,  автоматизации технологи...       От 3 до 6 лет  3702
4  Интернет-компания (поисковики, платежные систе...  От 1 года до 3 лет  2870
5  Интернет-компания (поисковики, платежные систе...           Нет опыта  1930
6                                               Банк  От 1 года до 3 лет  1677
7  Интернет-компания (поисковики, платежные систе...       От 3 до 6 лет  1514
8                Разработка программного обеспечения           Нет опыта  1225
9  Системная интеграция,  автоматизации технологи...           Нет опыта  1217

# Выводы:

Наибольшее кол-во вакансий в Москве. Так же стоит отметить, что агрегатор работает не только в России, но и еще зарубежом, судя по присутствующим городам. Примерно у половины вакансий не заполнены сведения о заработной плате. Скорее всего это связано с потребностью работотаделей в гибкости при определении ЗП для каждого кандидата. Средняя нижняя граница ЗП составляет 71065, а верхняя 110537. Эти показатели ЗП соответствуют средему уровню ЗП IT специалиста. Также ранее был сделан вывод, что агрегатор имеет большую популярность у IT компаний, соответсвенно средние показатели ЗП на агрегаторе вакансий примерно соответствуют средним ЗП в IT индустрии. Наиболее популяным сочетанием "тип занятости"-"тип рабочего графика" с внушительным отрывом в 5 раз является "полная занятость" - "полный день" с более чем 35000 вакансиями. Самое частовстречаемое значение показателя "Опыт работы" является "От 1 года до 3 лет" с 26152 вакансиями. Также на агрегаторе больше всего junior/middle IT-вакансий.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [28]:
# текст запроса
query_5_1 = f'''
(select 
    e.name,
    count(*) as cnt_vac
from employers e
join vacancies v on v.employer_id = e.id
group by 1
order by 2 desc
limit 1)  -- получение 1-го работодателя по кол-ву вакансий

union all 

(select 
    e.name,
    count(*) as cnt_vac
from employers e
join vacancies v on v.employer_id = e.id
group by 1
order by 2 desc
offset 4 limit 1)  -- получение 5-го работодателя по кол-ву вакансий
'''

In [29]:
# результат запроса
df_5_1 = pd.read_sql_query(query_5_1, connection)
df_5_1

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\1147638897.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  cnt_vac
0         Яндекс     1933
1  Газпром нефть      331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [30]:
# текст запроса
query_5_2 = f'''
select
    a.name as name,
    count(e.id) as cnt_emp,
    count(v.id) as cnt_vac
from areas a
left join employers e on a.id = e.area  -- присоединение работодателей по региону
left join vacancies v on a.id = v.area_id  -- присоединение вакансий по региону
group by 1
having count(v.id) = 0  -- фильтр для регионов, где нет вакансий
order by 2 desc  -- сортировка для нахождения региона, в котором наибольшее кол-во работодателей
limit 1
'''

In [31]:
# результат запроса
df_5_2 = pd.read_sql_query(query_5_2, connection)
df_5_2

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\2306256317.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  cnt_emp  cnt_vac
0  Россия      410        0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [32]:
# текст запроса
query_5_3 = f'''
select
    e.name,
    count(distinct v.area_id) as cnt_areas  -- считаем кол-во уникальных регионов, для каждого работодателя
from employers e
join vacancies v on e.id = v.employer_id
group by 1
order by 2 desc
'''

In [33]:
# результат запроса
df_5_3 = pd.read_sql_query(query_5_3, connection)
df_5_3

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\133400711.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  cnt_areas
0                      Яндекс        181
1                  Ростелеком        152
2                  Спецремонт        116
3      Поляков Денис Иванович         88
4                   ООО ЕФИН          71
...                       ...        ...
14761                  UniSol          1
14762            UNISTORY LLC          1
14763                   UNIT6          1
14764     United Distribution          1
14765              UNITEL LLC          1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [34]:
# текст запроса
query_5_4 = f'''
select
    count(*) as cnt
from employers e
left join employers_industries ei on e.id = ei.employer_id
left join industries i on ei.industry_id = i.id
where i.id is null  -- фильтр по работадателям, у которых не указана сфера деятельности
'''

In [35]:
# результат запроса
df_5_4 = pd.read_sql_query(query_5_4, connection)
df_5_4

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\1045119256.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt
0  8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [36]:
# текст запроса
query_5_5 = f'''
select
    e.name,
    count(i.id) as cnt  -- кол-во сфер деятельности компании
from employers e
left join employers_industries ei on e.id = ei.employer_id
left join industries i on ei.industry_id = i.id
group by 1
having count(i.id) = 4  -- фильтр для компаний с 4 сферами деятельности
order by 1
offset 2 limit 1  -- получение 3-го по алфавиту работодателя с 4-мя сферами деятельности 
'''

In [37]:
# результат запроса
df_5_5 = pd.read_sql_query(query_5_5, connection)
df_5_5

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\4025176131.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  cnt
0  2ГИС    4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [38]:
# текст запроса
query_5_6 = f'''
select
    count(*)
from employers e
left join employers_industries ei on e.id = ei.employer_id
left join industries i on ei.industry_id = i.id
where i.name = 'Разработка программного обеспечения'
'''

In [39]:
# результат запроса
df_5_6 = pd.read_sql_query(query_5_6, connection)
df_5_6

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\2713946032.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [40]:
# код для получения списка городов-милионников
import requests 
from bs4 import BeautifulSoup


url = "https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8"
response = requests.get(url)  

page = BeautifulSoup(response.text, "html.parser")
table = page.find('table')  # получение таблицы с регонами-миллионниками

cities = []

# цикл по всем строкам и столбцам таблицы с регонами-миллионниками
for row in table.find_all('tr'):
    columns = row.find_all('td')
    if len(columns) > 1:  # условие на непустую ячейку
        cities.append(columns[1].text)

cities = tuple(cities)
print(cities)

('Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург', 'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск', 'Уфа', 'Самара', 'Ростов-на-Дону', 'Краснодар', 'Омск', 'Воронеж', 'Пермь', 'Волгоград')


In [41]:
# текст запроса
query_5_7 = f'''
(select
    a.name,
    count(v.id) as cnt
from areas a
left join vacancies v on v.area_id = a.id
left join employers_industries ei on ei.employer_id = v.employer_id
left join industries i on ei.industry_id = i.id
left join employers e on ei.employer_id = e.id
where e.name = 'Яндекс' and a.name in {cities}
group by a.name
order by 2)  --кол-во вакансий Яндекс для каждого города-миллионника

union all

(select 
    'Total' as name,
    count(*)
from areas a
left join vacancies v on v.area_id = a.id
left join employers_industries ei on ei.employer_id = v.employer_id
left join industries i on ei.industry_id = i.id
left join employers e on ei.employer_id = e.id
where e.name = 'Яндекс' and a.name in {cities})  -- подсчет общего кол-ва вакансий Яндекс в городах миллониках
order by 2
'''

In [42]:
# результат запроса
df_5_7 = pd.read_sql_query(query_5_7, connection)
df_5_7

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\2595257991.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  cnt
0              Омск   21
1         Челябинск   22
2        Красноярск   23
3         Волгоград   24
4    Ростов-на-Дону   25
5             Пермь   25
6            Казань   25
7               Уфа   26
8            Самара   26
9         Краснодар   30
10          Воронеж   32
11      Новосибирск   35
12  Нижний Новгород   36
13     Екатеринбург   39
14  Санкт-Петербург   42
15           Москва   54
16            Total  485

***

## Дополнительное исследование

Распределение кол-ва вакансий по работодателям (топ 10)

In [43]:
# текст запроса
query_5_8 = f'''
select 
    e.name,
    count(*) as cnt_vac
from employers e
join vacancies v on v.employer_id = e.id
group by 1
order by 2 desc
limit 10
'''

In [44]:
# результат запроса
df_5_8 = pd.read_sql_query(query_5_8, connection)
df_5_8

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\1966148404.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  cnt_vac
0                 Яндекс     1933
1             Ростелеком      491
2               Тинькофф      444
3                   СБЕР      428
4          Газпром нефть      331
5            ИК СИБИНТЕК      327
6                    МТС      292
7                DataArt      247
8  Совкомбанк Технологии      204
9             Первый Бит      176

Визуализация запроса

In [45]:
fig = px.bar(
    data_frame=df_5_8,
    x='name', 
    y='cnt_vac', 
    color='name', 
    text='cnt_vac', 
    orientation='v',
    height=500, 
    width=1000, 
    title='Топ 10 работодателей по вакансиям'
    )

fig.show()
fig.write_image('graphs/img3.png')
fig.write_html("html/img3.html")

![Топ 10 работодателей по вакансиям](graphs/img3.png)

Средняя ЗП по работодателям

In [46]:
# текст запроса
query_5_9 = f'''
select 
    e.name,
    round(avg(coalesce(
        (salary_from + salary_to) / 2,  -- значение ЗП, если есть оба поля зарплат 
        salary_from,  -- значение ЗП, если есть только нижняя граница зарплат
        salary_to,  -- значение ЗП, если есть только верхняя граница зарплат
        0
    ))) as avg_salary  -- расчет среднего значения для ЗП
from employers e
join vacancies v on v.employer_id = e.id
where e.name in  -- выборка из топ 10 работодателей по кол-ву вакансий
    (
        select name
        from (
                select 
                    e.name,
                    count(*) as cnt_vac
                from employers e
                join vacancies v on v.employer_id = e.id
                group by 1
                order by 2 desc
                limit 10
            ) as top_10_emp  -- топ 10 работодателей по 
    )
group by 1
order by 2 desc
limit 10
'''

In [47]:
# результат запроса
df_5_9 = pd.read_sql_query(query_5_9, connection)
df_5_9

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\2985122994.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  avg_salary
0             Первый Бит     69243.0
1             Ростелеком     37508.0
2                    МТС     20328.0
3                 Яндекс     20019.0
4                   СБЕР     11143.0
5            ИК СИБИНТЕК     11112.0
6  Совкомбанк Технологии      3573.0
7               Тинькофф       791.0
8          Газпром нефть       338.0
9                DataArt         0.0

Визуализация запроса

In [48]:
fig = px.bar(
    data_frame=df_5_9,
    x='name', 
    y='avg_salary', 
    color='name', 
    text='avg_salary', 
    orientation='v',
    height=500, 
    width=1000, 
    title='Средняя ЗП по работодателям по самым активным работодателям'
    )

fig.show()
fig.write_image('graphs/img4.png')
fig.write_html("html/img4.html")

![Средняя ЗП по работодателям по самым активным работодателям](graphs/img4.png)

Зависимость средней ЗП от кол-ва вакансий работодателя

In [49]:
# текст запроса
query_5_10 = f'''
select 
    e.name,
    count(*) as cnt_vac,
    round(avg(coalesce(
        (salary_from + salary_to) / 2,  -- значение ЗП, если есть оба поля зарплат 
        salary_from,  -- значение ЗП, если есть только нижняя граница зарплат
        salary_to,  -- значение ЗП, если есть только верхняя граница зарплат
        0
    ))) as avg_salary  -- расчет среднего значения для ЗП
from employers e
join vacancies v on v.employer_id = e.id
where e.name in  -- выборка из топ 10 работодателей по кол-ву вакансий
    (
        select name
        from (
                select 
                    e.name,
                    count(*) as cnt_vac
                from employers e
                join vacancies v on v.employer_id = e.id
                group by 1
                order by 2 desc
                limit 100
            ) as top_10_emp  -- топ 10 работодателей по 
    )
group by 1
order by 2 desc
limit 100
'''

In [50]:
# результат запроса
df_5_10 = pd.read_sql_query(query_5_10, connection)
df_5_10

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\3049816343.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name  cnt_vac  avg_salary
0                 Яндекс     1933     20019.0
1             Ростелеком      491     37508.0
2               Тинькофф      444       791.0
3                   СБЕР      428     11143.0
4          Газпром нефть      331       338.0
..                   ...      ...         ...
95             ISPsystem       45      6667.0
96                  Эр-1       44         0.0
97           Ресурс, ГАП       43     11169.0
98  LeverX International       43         0.0
99              inDriver       42     22545.0

[100 rows x 3 columns]

Визуализация запроса

In [51]:
fig_scatter = px.scatter(
    data_frame=df_5_10,
    x='avg_salary',
    y='cnt_vac',
    color='name',
    title='Зависимость ЗП от кол-во вакансий для вакансий топ-10 работодателей')

fig_scatter.show()
fig_scatter.write_image('graphs/img5.png')
fig_scatter.write_html("html/img5.html")

![Зависимость ЗП от кол-во вакансий для вакансий топ-10 работодателей](graphs/img5.png)

Количество работодателей и вакансий для каждого региона

In [52]:
# текст запроса
query_5_11 = f'''
select
    a.name as name,
    count(e.id) as cnt_emp,
    count(v.id) as cnt_vac
from areas a
left join employers e on a.id = e.area  -- присоединение работодателей по региону
left join vacancies v on a.id = v.area_id  -- присоединение вакансий по региону
group by 1
order by 2 desc  -- сортировка для нахождения региона, в котором наибольшее кол-во работодателей
limit 10
'''

In [53]:
# результат запроса
df_5_11 = pd.read_sql_query(query_5_11, connection)
df_5_11

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\1289296352.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



name   cnt_emp   cnt_vac
0           Москва  31272712  31272712
1  Санкт-Петербург   6320667   6320667
2            Минск   2354880   2354880
3           Алматы   1364132   1364132
4      Новосибирск   1149438   1149438
5     Екатеринбург   1034082   1034082
6  Нижний Новгород    711420    711420
7           Казань    679200    679200
8        Краснодар    532109    532109
9   Ростов-на-Дону    378885    378885

# Выводы:

В десятку работодателей по кол-ву вакансий входят самые известные представители финтеха такие, как СБЕР, Тинькоф, Совкомбанк и т.д. Также в этом топе присутствует большое кол-во IT-компаний, одно из которых является лидером по кол-ву вакансий, это Яндекс. По кол-ву работодателей и вакансий выделяются города миллионники, что логично. В топ работодателей по кол-ву регионов с вакансиями, также отличился Яндекс, также стоит отметить необычного работодателя "Поляков Денис Иванович", который разместил свои вакансии в 88 регионах, возможно мошенничество. Более 8 тысяч из 23 тысяч работодателей не указывают сферу деятлеьности. Для Яндекса среднее кол-во вакансий в городах миллиониках РФ примерно 30. Средняя ЗП на вакансиях работодателя зависит от кол-ва этих вакансий, чем больше вакансий тем ниже ЗП (возможно выбросы из-за того, что работодатели не указывают зарплатную вилку). Самая высокая средняя ЗП среди топ-10 работодателей по кол-ву вакансий у Первый бит и она составляет 69243 рубля.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [54]:
# текст запроса
query_6_1 = f'''
select
    count(*) as cnt
from vacancies v
where lower(v.name) like '%data%' or lower(v.name) like '%данн%'
'''

In [55]:
# результат запроса
df_6_1 = pd.read_sql_query(query_6_1, connection)
df_6_1

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\1281472032.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt
0  1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [56]:
# текст запроса
query_6_2 = f'''
select
    count(*) as cnt
from (
        select *
        from vacancies
        where (lower(name) like '%junior%')
        or (experience = 'Нет опыта')
        or (employment = 'Стажировка')
    ) as jun_v  -- подзапрос для выбора только Junior вакансий
where
    (name ilike '%data scientist%')
    or (name ilike '%data science%')
    or (name ilike '%исследователь данных%')
    or (name ilike '%ML%' and name not ilike '%HTML%')
    or (name ilike '%machine learning%')
    or (name ilike '%машинн%обучен%')  -- условие для определения, что вакансия для DS
'''

In [57]:
# результат запроса
df_6_2 = pd.read_sql_query(query_6_2, connection)
df_6_2

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\118162426.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



cnt
0   51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [58]:
# текст запроса
query_6_3 = f'''
select 
    count(*)
from vacancies
where
    (
        (name ilike '%data scientist%')
        or (name ilike '%data science%')
        or (name ilike '%исследователь данных%')
        or (name ilike '%ML%' and name not ilike '%HTML%')
        or (name ilike '%machine learning%')
        or (name ilike '%машинн%обучен%')
    )  -- условие для определения, что вакансия для DS
    and 
    (
        key_skills like '%SQL%' 
        or key_skills like '%postgres%'
    )  -- условие на наличие SQL и Postgre в ключевых навыках DS
'''

In [59]:
# результат запроса
df_6_3 = pd.read_sql_query(query_6_3, connection)
df_6_3

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\3310962512.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0    229

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [60]:
# текст запроса
query_6_4 = f'''
select 
    count(*)
from vacancies
where
    (
        (name ilike '%data scientist%')
        or (name ilike '%data science%')
        or (name ilike '%исследователь данных%')
        or (name ilike '%ML%' and name not ilike '%HTML%')
        or (name ilike '%machine learning%')
        or (name ilike '%машинн%обучен%')
    )  -- условие для определения, что вакансия для DS
    and 
    (
        key_skills like '%Python%'
    )  -- условие на наличие Python в ключевых навыках DS
'''

In [61]:
# результат запроса
df_6_4 = pd.read_sql_query(query_6_4, connection)
df_6_4

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\614304167.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0    357

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [62]:
# текст запроса
query_6_5 = f'''
select 
    round(
        avg(
            length(key_skills) - length(replace(key_skills, chr(9), '')) + 1
        ), 2) as avg_skills  -- подсчет среднего кол-во ключевых навыков в вакансих для DS
from vacancies
where
    (
        (name ilike '%data scientist%')
        or (name ilike '%data science%')
        or (name ilike '%исследователь данных%')
        or (name like '%ML%' and name not ilike '%HTML%')
        or (name ilike '%machine learning%')
        or (name ilike '%машинн%обучен%')
    )  -- условие для определения, что вакансия для DS
    and 
    (
        key_skills is not null
    )  -- условие, что поле ключевые навыки не пустое
'''

In [63]:
# результат запроса
df_6_5 = pd.read_sql_query(query_6_5, connection)
df_6_5

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\4089566367.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



avg_skills
0        6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [64]:
# текст запроса
query_6_6 = f'''
select 
    round(avg(coalesce(
        (salary_from + salary_to) / 2,  -- значение ЗП, если есть оба поля зарплат 
        salary_from,  -- значение ЗП, если есть только нижняя граница зарплат
        salary_to,  -- значение ЗП, если есть только верхняя граница зарплат
        0
    ))) as avg_salary,  -- расчет среднего значения для ЗП
    experience
from vacancies
where
    (
        (name ilike '%data scientist%')
        or (name ilike '%data science%')
        or (name ilike '%исследователь данных%')
        or (name like '%ML%' and name not like '%HTML%')
        or (name ilike '%machine learning%')
        or (name ilike '%машинн%обучен%')
    )  -- условие для определения, что вакансия для DS
    and
    (
        (salary_from is not null) or (salary_to is not null)
    )  -- условие, что у вакансии будет хотя бы одно поле для зарплат
group by 2
order by 1
'''

In [65]:
# результат запроса
df_6_6 = pd.read_sql_query(query_6_6, connection)
df_6_6

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\2297777544.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



avg_salary          experience
0     74643.0           Нет опыта
1    139675.0  От 1 года до 3 лет
2    243115.0       От 3 до 6 лет

***

## Дополнительное исследование

Общее кол-во вакансий для DS

In [66]:
# текст запроса
query_6_7 = f'''
select 
    count(*)
from vacancies
where
    (
        (name ilike '%data scientist%')
        or (name ilike '%data science%')
        or (name ilike '%исследователь данных%')
        or (name ilike '%ML%' and name not ilike '%HTML%')
        or (name ilike '%machine learning%')
        or (name ilike '%машинн%обучен%')
    )  -- условие для определения, что вакансия для DS
'''

In [67]:
# результат запроса
df_6_7 = pd.read_sql_query(query_6_7, connection)
df_6_7

C:\Users\alpavlov\AppData\Local\Temp\ipykernel_10984\3777273199.py:2: UserWarning:

pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.



count
0    536

# Выводы:

Всего на площадке предствалено 1771 вакансий связанных с данными, из которых 536 связанны с DataScience. На агрегаторе также представлено 51 вакансия для junior DS специалистов. Средняя ЗП DS без опыта - 74643 рублей, от 1 года до 3 лет - 139675 рублей, от 3 до 6 лет - 243115 рублей. Так же для DS в среднем указывают чуть больше 6 ключевых навыка. Одними из самых популярных навыков являются вледения SQL и Python.

# Общий вывод по проекту

В базе данных агрегатора вакансий представлено 49197 вакансий, 23501 запись о работодателях, данные о 1362 регионах и 294 сферах деятельности работодателей. На одного работодателя в среднем приходится примерно две вакансии. Большинство вакансий относится к сфере разработки программного обеспечения и IT, что может указывать на ориентированность агрегатора на IT-сектор. На платформе представлены 5 типов рабочих графиков и 4 типа занятости.

Наибольшее количество вакансий сосредоточено в Москве. Агрегатор работает не только в России, но и за рубежом. У половины вакансий отсутствуют данные о заработной плате, что может свидетельствовать о гибкости работодателей в определении ЗП. Средняя нижняя граница ЗП составляет 71065 рублей, а верхняя – 110537 рублей, что соответствует среднему уровню ЗП IT-специалистов. Наиболее популярным сочетанием "тип занятости" - "тип рабочего графика" является "полная занятость" - "полный день" с более чем 35000 вакансиями. Самое часто встречающееся значение показателя "Опыт работы" – "От 1 года до 3 лет" с 26152 вакансиями, что указывает на преобладание junior/middle IT-вакансий.

В десятку работодателей по количеству вакансий входят крупные представители финтеха (СБЕР, Тинькоф, Совкомбанк и т.д.) и IT-компании, такие как Яндекс, который лидирует по количеству вакансий. Вакансии сконцентрированы в городах-миллионниках. Яндекс также лидирует по количеству регионов с вакансиями. Необычным работодателем является "Поляков Денис Иванович", разместивший вакансии в 88 регионах, что может свидетельствовать о возможном мошенничестве. Более 8000 из 23000 работодателей не указывают сферу деятельности. Средняя ЗП по вакансиям работодателя зависит от количества вакансий: чем больше вакансий, тем ниже средняя ЗП. Самая высокая средняя ЗП среди топ-10 работодателей у "Первый бит" – 69243 рубля.

На платформе представлено 1771 вакансия, связанная с данными, из них 536 вакансий относятся к DataScience. 51 вакансия предназначена для junior DS специалистов. Средняя ЗП DS специалистов без опыта – 74643 рублей, от 1 до 3 лет – 139675 рублей, от 3 до 6 лет – 243115 рублей. В вакансиях для DS специалистов в среднем указывают чуть больше 6 ключевых навыков. Среди самых популярных навыков – владение SQL и Python. Эти обобщенные выводы позволяют понять основные тенденции и особенности вакансий и работодателей на агрегаторе, а также уровень заработной платы и требования к ключевым навыкам в сфере DataScience.